pkg/services/notary/notary.go

```diff
@@ -0,0 +1,376 @@
+	if validationErr == nil {
+	loop:
+		for i, w := range payload.MainTransaction.Scripts {
+			if payload.MainTransaction.Signers[i].Account.Equals(n.Config.Chain.GetNotaryContractScriptHash()) {
+				continue
+			}
+			if len(w.VerificationScript) != 0 {
+				switch r.typ {
+				case Signature:
+					if !exists {
+						r.nSigsCollected++
+					} else if len(r.main.Scripts[i].VerificationScript) == 0 { // need this check because signature can already be added (consider receiving the same payload multiple times)
+						r.main.Scripts[i] = w
+						r.nSigsCollected++
+					}
+					if r.nSigsCollected == r.nSigs {
+						break loop
+					}
+				case MultiSignature:
+					if r.sigs == nil {
+						r.sigs = make(map[*keys.PublicKey][]byte)
+					}
+
+					hash := r.main.GetSignedHash().BytesBE()
+					for _, pub := range pubs {
+						if r.sigs[pub] != nil {
+							continue // signature for this pub has already been added
+						}
+						if pub.Verify(w.InvocationScript[2:], hash) { // then pub is the owner of the signature
+							r.sigs[pub] = w.InvocationScript
+							r.nSigsCollected++
+							if r.nSigsCollected == r.nSigs {
+								var invScript []byte
+								for j := range pubs {
+									if sig, ok := r.sigs[pubs[j]]; ok {
+										invScript = append(invScript, sig...)
+									}
+								}
+								r.main.Scripts[i].InvocationScript = invScript
+							}
+							break loop
+						}
+					}
+					// pubKey was not found for the signature i.e. signature is bad - we're OK with that, let the fallback TX to be added
+					break loop // only one multisignature is allowed
+				}
+			}
+		}
+	}
+	if r.typ != Unknown && r.nSigsCollected == nSigs && r.minNotValidBefore > n.Config.Chain.BlockHeight() {
+		if err := n.finalize(r.main); err != nil {
+			n.Config.Log.Error("failed to finalize main transaction", zap.Error(err))
+		} else {
+			r.isSent = true
+		}
+	}
+}
+
+// OnRequestRemoval is a callback which is called after fallback transaction is removed
+// from the notary payload pool due to expiration, main tx appliance or any other reason.
+func (n *Notary) OnRequestRemoval(pld *payload.P2PNotaryRequest) {
+	if n.getAccount() == nil {
+		return
+	}
+
+	n.reqMtx.Lock()
+	defer n.reqMtx.Unlock()
+	r, ok := n.requests[pld.MainTransaction.Hash()]
+	if !ok {
+		return
+	}
+	for i, fb := range r.fallbacks {
+		if fb.Hash().Equals(pld.FallbackTransaction.Hash()) {
+			r.fallbacks = append(r.fallbacks[:i], r.fallbacks[i+1:]...)
+			break
+		}
+	}
+	if len(r.fallbacks) == 0 {
+		delete(n.requests, r.main.Hash())
+	}
+}
+
+// PostPersist is a callback which is called after new block is persisted.
+func (n *Notary) PostPersist(bc blockchainer.Blockchainer, pool *mempool.Pool, b *block.Block) {
+	if n.getAccount() == nil {
+		return
+	}
+
+	n.reqMtx.Lock()
+	defer n.reqMtx.Unlock()
+	for h, r := range n.requests {
+		if !r.isSent && r.typ != Unknown && r.nSigs == r.nSigsCollected && r.minNotValidBefore > bc.BlockHeight() {
```

Comment:
I think we need a flag for `fallbackSent`. Consider 5/7 multisig, you're receiving 2 signatures before NVB kicks in, then you've sent 2 fallbacks and suddenly you receive 5 other requests. Even though you technically can collect 5/7 multisig the main transaction will never get in because it's conflicting with fallbacks you've already sent.

Reply:
But we have `minNotValidBefore`, and if `minNotValidBefore` kicks in, then we won't ever try to send the main transaction. So `minNotValidBefore` replaces `fallbackSent` flag. Am I missing something?

Reply:
No, you're right, for some reason I thought it's moving as we start sending fallbacks, but it's not, so everything is fine here.